README
Games-Domino

The README  is  used  to  introduce the module and provide instructions on how to
install the module, any machine dependencies it may have (for example C compilers
and installed libraries) and any other information that should be provided before
the module is installed.

A  README  file  is required for CPAN modules since CPAN extracts the README file
from a module distribution so that people browsing the archive can use it to  get
an  idea  of  the  module's  uses.  It  is usually a good idea to provide version
information here so that people can decide whether fixes for the module are worth
downloading.

INSTALLATION

To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install

SUPPORT AND DOCUMENTATION

After  installing,   you can  find documentation for this module with the perldoc
command.

    perldoc Games::Domino

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=Games-Domino

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/Games-Domino

    CPAN Ratings
        http://cpanratings.perl.org/d/Games-Domino

    Search CPAN
        http://search.cpan.org/dist/Games-Domino/

LICENSE AND COPYRIGHT

Copyright (C) 2012 - 2015 Mohammad S Anwar

This program is free software; you can redistribute it and/or modify it under the
terms of the the Artistic License (2.0).You may obtain a copy of the full license
at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying / distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you, you're nevertheless required to ensure that your Modified Version
complies with the requirements of this license.

This  license  does  not  grant you the right to use any trademark, service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to  make,  have made, use, offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by the  Package.  If you institute patent litigation
(including  a  cross-claim or  counterclaim)  against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF     MERCHANTABILITY,   FITNESS   FOR   A PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL, OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE,EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

Changes
Revision history for Games::Domino

0.29    Wed Apr 26 13:55:00 2017
        - Exposed 'cheat' flag.

0.28    Tue Apr 25 10:15:00 2017
        - Updated game script to be more user friendly.

0.27    Sun Apr 23 10:50:00 2017
        - Updated how the instructions get displayed.
        - Updated result to be more descriptive.

0.26    Mon Apr 03 16:40:00 2017
        - Added test scripts (t/meta-json.t and t/meta-yml.t).

0.25    Wed Oct 05 10:40:00 2016
        - Documented method reset() in the package Games::Domino::Player.
        - Added fallback switch to overload pragma.

0.24    Thu Sep 29 10:50:00 2016
        - Improved type definitions for 'ZeroOrOne', 'ZeroToSix' and 'HorC'.

0.23    Wed Aug 24 10:10:00 2016
        - Updated copyright year information.
        - Tidied up unit test scripts.

0.22    Wed Mar 09 15:45:00 2016
        - Assigned minimum version required for Moo and namespace::clean.
        - Added key 'provides' to the Makefile.PL script.

0.21    Sun Sep 27 09:40:00 2015
        - Updated section "SYNOPSIS" to reflect the changes to the game script.

0.20    Fri Jun 05 12:00:00 2015
        - Fixed method result() to spit out the right result.
        - Dropped private method _play() and moved the code to the method play().
        - Updated the game script to respect the above changes.
        - Updated the section "SYNOPSIS" w.r.t the changes to the game script.

0.19    Thu Jun 04 12:00:00 2015
        - Updated method result() to return correct result string efficiently.
        - Refactored method show() to return state of the game.
        - Added new method instructions().
        - Updated game script to respect the above changes.

0.18    Wed Jun 03 10:35:00 2015
        - Updated method result() to return the result.
        - Updated method play() to continue unless the game is over.
        - Updated method show() to show 'EMPTY' if the game board is empty.

0.17    Mon Jun 01 13:35:00 2015
        - Added dependency on Term::ANSIColor::Markup in the Makefile.PL script.
        - Added new attribute 'color' to the class Games::Domino::Tile.
        - Updated method _play() to mark with color code who picked the tile.
        - Updated method _board() to return color coded board.

0.16    Fri May 29 10:20:00 2015
        - Fixed the failed Kwalitee Core metric 'consistent version'.

0.15    Wed May 27 09:50:00 2015
        - Added method get_available_tiles()
        - Updated method is_valid_tile()
        - Updated game script w.r.t the above addition.
        - Updated code in the section "SYNOPSIS".

0.14    Tue May 26 14:50:00 2015
        - Updated method play() and extracted out prompting for data.
        - Added method is_valid_tile()
        - Updated method _play() w.r.t above changes.
        - Added method _pick_from_bank()
        - Refactored method is_over()
        - Refactored method _next()
        - Updated game script w.r.t the above changes.

0.13    Fri May 22 09:55:00 2015
        - Added method reset() to the package Games::Domino.
        - Added method reset() to the package Games::Domino::Player.
        - Updated game script to allow replay the game in the same session.
        - Tidied up pod document of the methods in the package Games::Domino.

0.12    Wed May 13 11:00:00 2015
        - Fine-tuned the flow of the game.
        - Improved method is_over() slightly.
        - Renamed the script play-domino.
        - Updated pod document to reflect the above changes.

0.11    Mon Mar 23 12:50:00 2015
        - Updated Makefile.PL script to install the game script alongwith the package.
        - Updated pod document about the same.

0.10    Thu Feb 19 15:50:50 2015
        - Fixed the typo in the pod document of Games::Domino.

0.09    Thu Feb 19 14:10:50 2015
        - Added section SYNOPSIS.
        - Minor tweak to the code in the pod document.

0.08    Thu Feb 05 11:10:50 2015
        - Updated MANIFEST.SKIP file to ignore folder '_eumm'.
        - Added LICENSE file.
        - Defined $AUTHORITY in all the packages of the distribution.
        - Updated Copyright year information.

0.07    Fri Dec 12 10:36:50 2014
        - Tidied up the pod document.

0.06    Fri Dec 12 03:00:50 2014
        - Re-structured the code using 'Moo' i.e. downgrading from 'Moose'.
        - Added package Games::Domino::Params to support common data types.
        - Deleted Build.PL file.
        - Deleted SIGNATURE file.
        - Updated unit test to respect 'Moo'.

0.05    Mon Aug 20 13:11:50 2012
        - Fixed the typo in the pod for "DESCRIPTION".

0.04    Sat Aug 18 21:33:29 2012
        - Updated the text for "DESCRIPTION".

0.03    Sat Aug 18 18:19:20 2012
        - Major changes to let the user play with the computer (interactive mode).
        - Fixed couple of bugs encountered while making changes.
        - Minor cosmetics changes.
        - Updated the sample game script (play-domino.pl) to reflect the changes in the core module.

0.02    Thu Jul 26 16:38:29 2012
        - Extended Games::Domino::Tile class to hold double property
          as well top/bottom. This is to support the vertical branch.
        - Added unit test around the new attributes.
        - Fixed few typo in the pod for Games::Domino::Tile class.

0.01    Thu Jul 26 04:00:00 2012
        - First version, released.

lib/Games/Domino.pm
package Games::Domino;

$Games::Domino::VERSION   = '0.29';
$Games::Domino::AUTHORITY = 'cpan:MANWAR';

=head1 NAME

Games::Domino - Interface to the Domino game.

=head1 VERSION

Version 0.29

=cut

use 5.006;
use Data::Dumper;
use Term::ReadKey;
use Term::Screen::Lite;
use List::Util qw(shuffle);
use Term::ANSIColor::Markup;

use Games::Domino::Tile;
use Games::Domino::Player;
use Games::Domino::Params qw(ZeroOrOne ZeroToSix);

use Moo;
use namespace::clean;

use overload q{""} => 'as_string', fallback => 1;

has 'stock'    => (is => 'rw');
has 'board'    => (is => 'rw');
has 'human'    => (is => 'rw');
has 'computer' => (is => 'rw');
has 'current'  => (is => 'rw');
has 'board_l'  => (is => 'rw', isa => ZeroToSix);
has 'board_r'  => (is => 'rw', isa => ZeroToSix);
has 'cheat'    => (is => 'ro', isa => ZeroOrOne, default => sub { 0 });
has 'debug'    => (is => 'rw', isa => ZeroOrOne, default => sub { 0 });
has 'action'   => (is => 'rw', default => sub { [] });
has 'screen'   => (is => 'ro', default => sub { Term::Screen::Lite->new; });

=head1 DESCRIPTION

This is a very basic Domino game played by two players (Computer vs Human) at the
moment. This is just an  initial draft  of  Proof of Concept, also to get my head
around the game which I have never played in my life before.There is a cheat flag
which makes tiles for "Computer" visible to the other player "Human".  Avoid this
flag if possible.By default the cheat flag is turned off.There is  verbose switch
as well which is turned off by default. They  are  arranged  like  here before we
shuffle to start the the game.

    [0 | 0]
    [0 | 1] [1 | 1]
    [0 | 2] [1 | 2] [2 | 2]
    [0 | 3] [1 | 3] [2 | 3] [3 | 3]
    [0 | 4] [1 | 4] [2 | 4] [3 | 4] [4 | 4]
    [0 | 5] [1 | 5] [2 | 5] [3 | 5] [4 | 5] [5 | 5]
    [0 | 6] [1 | 6] [2 | 6] [3 | 6] [4 | 6] [5 | 6] [6 | 6]

=head1 SYNOPSIS

Below is the working code for the Domino game using the L<Games::Domino> package.
The game script C<play-domino> is supplied with the distribution and  on install,
is available to play with.

  USAGE: play-domino [-h] [long options...]

    --verbose  Play the game in verbose mode.
    --cheat    Play the game in cheat mode.

    --usage    show a short help message
    -h         show a compact help message
    --help     show a long help message
    --man      show the manual

=cut

sub BUILD {
    my ($self) = @_;

    $self->{human} = Games::Domino::Player->new({ name => 'H', show => 1 });
    if ($self->cheat) {
        $self->{computer} = Games::Domino::Player->new({ name => 'C', show => 1 });
    }
    else {
        $self->{computer} = Games::Domino::Player->new({ name => 'C' });
    }

    $self->_init;
}

=head1 METHODS

=head2 play()

Pick a tile from the current player. If no matching tile found then picks it from
the stock until it found  one or the stock has only 2 tiles left at that time the
game is over.

=cut

sub play {
    my ($self, $index) = @_;

    my $player = $self->current;
    my $name   = $player->name;

    if (defined $index) {
        if ($index =~ /^B$/i) {
            $self->_pick_from_bank($player);
        }
        else {
            my $tile = $player->_tile($index);
            print {*STDOUT} "[H] [P]: $tile [S]\n" if $self->debug;
            splice(@{$player->{bank}}, $index-1, 1);
            $self->_save($tile);
        }
    }
    else {
        my $tile = $player->pick($self->board_l, $self->board_r);
        if (defined $tile) {
            print {*STDOUT} "[C] [P]: $tile [S]\n" if $self->debug;
            $self->_save($tile);
        }
        else {
            $self->_pick_from_bank($player);
        }
    }
}

=head2 get_available_tiles()

Returns all available tile's index.

=cut

sub get_available_tiles {
    my ($self) = @_;

    return $self->current->_available_indexes;
}

=head2 is_valid_tile($index)

Return 1/0 depending on whether the tile at the given C<$index> is valid or not.

=cut

sub is_valid_tile {
    my ($self, $index) = @_;

    return (defined($index)
            && (($index =~ /^B$/i)
                ||
                ($self->current->_validate_index($index)
                 && $self->current->_validate_tile($index, $self->board_l, $self->board_r))
               ));
}

=head2 is_over()

Returns 1 or 0 depending whether the game is over or not.The game can be declared
over in the following circumstances:

=over 2

=item * Any one of the two players have used all his tiles.

=item * There are only two (2) tiles left in the bank.

=back

=cut

sub is_over {
    my ($self) = @_;

    return ((scalar(@{$self->{stock}}) == 2)
            ||
            (scalar(@{$self->{human}->{bank}}) == 0)
            ||
            (scalar(@{$self->{computer}->{bank}}) == 0));
}

=head2 result()

Declares who is the winner against whom and by how much margin.

=cut

sub result {
    my ($self) = @_;

    my ($result);
    if (scalar(@{$self->stock}) == 2) {
        my $c_b = scalar(@{$self->computer->bank});
        my $h_b = scalar(@{$self->human->bank});
        my $msg = 'Bank has only 2 tiles left. ';
        if ($c_b < $h_b) {
            $result = $self->_result("${msg}. Therefore, computer is declared the winner, having less tiles than you.", $c_b, $h_b);
        }
        elsif ($c_b > $h_b) {
            $result = $self->_result("${msg}. Therefore, you are declared the winner, having less tiles than computer.", $h_b, $c_b);
        }
        else {
            $result = $self->_result('${msg}. Therefore, game is declared draw as both players have the same number of tiles.', $h_b, $c_b);
        }
    }
    else {
        my $h = $self->human->value;
        my $c = $self->computer->value;

        if (scalar(@{$self->human->bank}) == 0) {
            $result = $self->_result('Congratulation, you are the winner as you have no tiles left.', $h, $c);
        }
        elsif (scalar(@{$self->computer->bank}) == 0) {
            $result = $self->_result('Sorry, computer is the winner as it has no tiles left.', $c, $h);
        }
        else {
            if ($h < $c) {
                $result = $self->_result('Congratulation, you are the winner as your remaining tiles value is less than computer.', $h, $c);
            }
            elsif ($h > $c) {
                $result = $self->_result('Sorry, computer is the winner as it\'s remaining tiles value is less than yours.', $c, $h);
            }
            else {
                $result = $self->_result('Game is declared draw as both the players reamaining tiles value is the same.', $c, $h);
            }
        }
    }

    if ($self->debug) {
        $result = sprintf("STOCK : %s\n%s", $self->as_string, $result);
    }

    return Term::ANSIColor::Markup->colorize($result);
}

=head2 show()

Returns the current state of the game.

=cut

sub show {
    my ($self) = @_;

    my $game = sprintf("%s\n", $self->_line);
    $game .= sprintf("[C]: %s\n", $self->computer->as_string);
    $game .= sprintf("[H]: %s\n", $self->human->as_string);
    $game .= "[G]: " . $self->_board . "\n";
    $game .= sprintf("%s", $self->_line);

    return $game;
}

=head2 reset()

Reset the game.

=cut

sub reset {
    my ($self) = @_;

    $self->human->reset;
    $self->computer->reset;
    $self->_init;
}

=head2 as_string()

Returns all the unused tiles remained in the bank.

=cut

sub as_string {
    my ($self) = @_;

    return '[EMPTY]' unless scalar(@{$self->{stock}});

    my $domino = '';
    foreach (@{$self->{stock}}) {
        $domino .= sprintf("%s==", $_);
    }

    $domino =~ s/[\=]+\s?$//;
    $domino =~ s/\s+$//;
    return $domino;
}

sub pause {
    my ($self, $message) = @_;

    $message = "Press any key to continue..." unless defined $message;
    print {*STDOUT} $message;

    $self->read_mode('cbreak');
    ReadKey(0);
    $self->read_mode;
}

sub read_mode {
    my ($self, $state) = @_;

    $state = 'normal' unless defined $state;
    ReadMode $state;
}

sub about_game {
    my ($self) = @_;

    return qq {
+-------------------------------------------------------------------------------+
|                                                                               |
|                              Games::Domino v$Games::Domino::VERSION                              |
|                                                                               |
+-------------------------------------------------------------------------------+
Tiles are numbered left to right starting with 1. Symbols used in this game are:
    [C]: Code for the computer player
    [H]: Code for the human player
    [P]: Personal tile
    [B]: Tile picked from the bank
    [S]: Successfully found the matching tile
    [F]: Unable to find the matching tile
    [G]: All matched tiles so far
+-------------------------------------------------------------------------------+
};
}

sub how_to_play {
    my ($self) = @_;

    return qq {
Example:

[C] [P]: [5 | 6] [S]
Computer picked the tile [5 | 6] from   his own collection and successfully found
the matching on board.

[H] [P]: [6 | 6] [S]
Human picked the tile [6 | 6] from his own collection and successfully found  the
matching on board.

[C] [B]: [2 | 6] [S]
Computer randomly picked the tile [2 | 6] from the bank and successfully found the
matching on board.

[C] [B]: [3 | 4] [F]
Computer randomly picked the tile [3 | 4] from the bank and but failed to find the
matching on board.

[H] [B]: [2 | 2] [S]
Human randomly picked the tile [2 | 2] from the bank and successfully  found  the
matching on board.

[H] [B]: [3 | 6] [F]
Human randomly picked the tile [3 | 6] from the bank and but failed to  find  the
matching on board.
+-------------------------------------------------------------------------------+
};
}

#
#
# PRIVATE METHODS

sub _pick_from_bank {
    my ($self, $player) = @_;

    my $name = $player->name;
    while (scalar(@{$self->{stock}}) > 2) {
        my $_tile = $self->_pick();
        $player->save($_tile);
        my $tile = $player->pick($self->board_l, $self->board_r);
        if (defined $tile) {
            print {*STDOUT} "[$name] [B]: $tile [S]\n" if $self->debug;
            $self->_save($tile);
            last;
        }
        else {
            print {*STDOUT} "[$name] [B]: $_tile [F]\n" if $self->debug;
        }
    }
}

sub _save {
    my ($self, $tile) = @_;

    if (!defined($self->{board}) || (scalar(@{$self->{board}}) == 0)) {
        push @{$self->{board}}, $tile;
        $self->{board_l} = $tile->left;
        $self->{board_r} = $tile->right;
        $self->_action($tile);
        $self->_next;
        return;
    }

    if ($self->{board_r} == $tile->left) {
        push @{$self->{board}}, $tile;
        $self->{board_r} = $tile->right;
        $self->_action($tile);
        $self->_next;
        return;

    }
    elsif ($self->{board_r} == $tile->right) {
        my $L = $tile->left;
        my $R = $tile->right;
        $tile->right($L);
        $tile->left($R);
        push @{$self->{board}}, $tile;
        $self->{board_r} = $L;
        $self->_action($tile);
        $self->_next;
        return;
    }

    if ($self->{board_l} == $tile->left) {
        my $L = $tile->left;
        my $R = $tile->right;
        $tile->right($L);
        $tile->left($R);
        unshift @{$self->{board}}, $tile;
        $self->{board_l} = $R;
        $self->_action($tile);
        $self->_next;
        return;
    }
    elsif ($self->{board_l} == $tile->right) {
        unshift @{$self->{board}}, $tile;
        $self->{board_l} = $tile->left;
        $self->_action($tile);
        $self->_next;
        return;
    }

    return;
}

sub _action {
    my ($self, $tile) = @_;

    if (defined $self->{action} && scalar(@{$self->{action}}) == 2) {
        foreach (@{$self->board}) {
            $_->color('blue');
        }
        $self->{action} = [ $tile ];
    }
    else {
        push @{$self->{action}}, $tile;
    }

    if ($self->current->name eq 'H') {
        $tile->color('green');
    }
    else {
        $tile->color('red');
    }
}

sub _board {
    my ($self) = @_;

    my $board = '';
    if (scalar(@{$self->board})) {
        foreach (@{$self->board}) {
            $board .= sprintf("<%s><bold>%s</bold><\/%s>==", $_->color, $_, $_->color);
        }
    }
    else {
        $board .= '<blue><bold>EMPTY</bold></blue>';
    }

    $board =~ s/[\=]+\s?$//;
    $board =~ s/\s+$//;
    return Term::ANSIColor::Markup->colorize($board);
}

sub _result {
    my ($self, $title, $a, $b) = @_;

    my $result = sprintf("<blue><bold>%s ", $title);
    $result .= sprintf("Final score [</bold></blue><green><bold>%d</bold></green>", $a);
    $result .= sprintf("<blue><bold>] against [</bold></blue><red><bold>%d</bold></red>", $b);
    $result .= '<blue><bold>].</bold></blue>';

    return $result;
}

sub _line {
    my ($self) = @_;

    return "="x81;
}

sub _init {
    my ($self) = @_;

    $self->{stock} = $self->_prepare();
    $self->{board} = [];
    $self->{human}->save($self->_pick)    for (1..7);
    $self->{computer}->save($self->_pick) for (1..7);
    $self->{current} = $self->{human};
}

sub _pick {
    my ($self) = @_;

    return shift @{$self->{stock}};
}

sub _prepare {
    my ($self) = @_;

    my $tiles = [];
    my $tile  = Games::Domino::Tile->new({ left => 0, right => 0, double => 1 });
    push @$tiles, $tile;
    foreach my $R (1..6) {
        my $L = 0;
        my $D = 0;
        while ($R >= $L) {
            ($R == $L)?($D = 1):($D = 0);
            push @$tiles, Games::Domino::Tile->new({ left => $L, right => $R, double => $D });
            $L++;
        }
    }

    $tiles = [shuffle @{$tiles}];
    return $tiles;
}

sub _next {
    my ($self) = @_;

    if ($self->current->name eq 'H') {
        $self->current($self->computer);
    }
    else {
        $self->current($self->human);
    }
}

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/Games-Domino>

=head1 BUGS

Please report any bugs or feature requests to C<bug-games-domino at rt.cpan.org>,
or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Games-Domino>.
I will be notified, and then you'll automatically be notified of progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Games::Domino

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Games-Domino>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Games-Domino>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Games-Domino>

=item * Search CPAN

L<http://search.cpan.org/dist/Games-Domino/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2012 - 2016 Mohammad S Anwar.

This program  is  free software; you can redistribute it and / or modify it under
the  terms  of the the Artistic License (2.0). You may obtain a  copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of Games::Domino


lib/Games/Domino/Tile.pm
package Games::Domino::Tile;

$Games::Domino::Tile::VERSION   = '0.29';
$Games::Domino::Tile::AUTHORITY = 'cpan:MANWAR';

=head1 NAME

Games::Domino::Tile - Represents the tile of the Domino game.

=head1 VERSION

Version 0.29

=cut

use 5.006;
use Data::Dumper;
use Games::Domino::Params qw(ZeroOrOne ZeroToSix);

use Moo;
use namespace::clean;

use overload q{""} => 'as_string', fallback => 1;

has 'left'   => (is => 'rw', isa => ZeroToSix, required => 1);
has 'right'  => (is => 'rw', isa => ZeroToSix, required => 1);
has 'double' => (is => 'ro', isa => ZeroOrOne, required => 1);
has 'top'    => (is => 'rw', isa => ZeroToSix);
has 'bottom' => (is => 'rw', isa => ZeroToSix);
has 'color'  => (is => 'rw', default => sub { 'blue' });

=head1 DESCRIPTION

It is used internally by L<Games::Domino>.

=cut

around BUILDARGS => sub {
    my $orig  = shift;
    my $class = shift;

    unless (exists $_[0]->{double}) {
        if (defined($_[0]->{left})
            && defined($_[0]->{right}) && ($_[0]->{left} == $_[0]->{right})) {
            $_[0]->{double} = 1;
            $_[0]->{top} = $_[0]->{bottom} = $_[0]->{left};
        }
        else {
            $_[0]->{double} = 0;
        }
    }

    die("ERROR: Invalid double attribute for the tile.\n")
        if (defined($_[0]->{left})
            && defined($_[0]->{right})
            && ( (($_[0]->{left} == $_[0]->{right}) && ($_[0]->{double} != 1))
                 ||
                 (($_[0]->{left} != $_[0]->{right}) && ($_[0]->{double} != 0))
               )
        );

    if ($_[0]->{double} == 1) {
        $_[0]->{top} = $_[0]->{bottom} = $_[0]->{left};
    }

    return $class->$orig(@_);
};

=head1 METHODS

=head2 value()

Returns the value of the tile i.e. sum of left and right bips.

    use strict; use warnings;
    use Games::Domino::Tile;

    my $tile = Games::Domino::Tile->new({ left => 1, right => 4 });
    print "Value of the tile is [" . $tile->value . "].\n";

=cut

sub value {
    my ($self) = @_;

    return ($self->{left} + $self->{right});
}

=head2 as_string()

Returns the tile object as string. This method is overloaded as string context.So
if we print the object then this method gets called.  You can explictly call this
method as well. Suppose the tile has 3 left pips and 6 right pips then this would
return it as [3 | 6].

    use strict; use warnings;
    use Games::Domino::Tile;

    my $tile = Games::Domino::Tile->new({ left => 1, right => 4 });
    print "The tile is $tile\n";
    # same as above
    print "The tile is " . $tile->as_string() . "\n";

=cut

sub as_string {
    my ($self) = @_;

    return sprintf("[%d | %d]", $self->left, $self->right);
}

=head1 AUTHOR

Mohammad S Anwar, C<< <mohammad.anwar at yahoo.com> >>

=head1 REPOSITORY

L<https://github.com/manwar/Games-Domino>

=head1 BUGS

Please report any bugs or feature requests to C<bug-games-domino at rt.cpan.org>,
or through the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Games-Domino>.
I will be notified, and then you'll automatically be notified of progress on your
bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Games::Domino::Tile

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Games-Domino>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Games-Domino>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Games-Domino>

=item * Search CPAN

L<http://search.cpan.org/dist/Games-Domino/>

=back

=head1 LICENSE AND COPYRIGHT

Copyright 2012 - 2016 Mohammad S Anwar.

This program  is  free software; you can redistribute it and / or modify it under
the  terms  of the the Artistic License (2.0). You may obtain a  copy of the full
license at:

L<http://www.perlfoundation.org/artistic_license_2_0>

Any  use,  modification, and distribution of the Standard or Modified Versions is
governed by this Artistic License.By using, modifying or distributing the Package,
you accept this license. Do not use, modify, or distribute the Package, if you do
not accept this license.

If your Modified Version has been derived from a Modified Version made by someone
other than you,you are nevertheless required to ensure that your Modified Version
 complies with the requirements of this license.

This  license  does  not grant you the right to use any trademark,  service mark,
tradename, or logo of the Copyright Holder.

This license includes the non-exclusive, worldwide, free-of-charge patent license
to make,  have made, use,  offer to sell, sell, import and otherwise transfer the
Package with respect to any patent claims licensable by the Copyright Holder that
are  necessarily  infringed  by  the  Package. If you institute patent litigation
(including  a  cross-claim  or  counterclaim) against any party alleging that the
Package constitutes direct or contributory patent infringement,then this Artistic
License to you shall terminate on the date that such litigation is filed.

Disclaimer  of  Warranty:  THE  PACKAGE  IS  PROVIDED BY THE COPYRIGHT HOLDER AND
CONTRIBUTORS  "AS IS'  AND WITHOUT ANY EXPRESS OR IMPLIED WARRANTIES. THE IMPLIED
WARRANTIES    OF   MERCHANTABILITY,   FITNESS   FOR   A   PARTICULAR  PURPOSE, OR
NON-INFRINGEMENT ARE DISCLAIMED TO THE EXTENT PERMITTED BY YOUR LOCAL LAW. UNLESS
REQUIRED BY LAW, NO COPYRIGHT HOLDER OR CONTRIBUTOR WILL BE LIABLE FOR ANY DIRECT,
INDIRECT, INCIDENTAL,  OR CONSEQUENTIAL DAMAGES ARISING IN ANY WAY OUT OF THE USE
OF THE PACKAGE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

=cut

1; # End of Games::Domino::Tile


Makefile.PL
#!/usr/bin/perl

use 5.006;
use strict; use warnings;
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME               => 'Games::Domino',
    AUTHOR             => q{Mohammad S Anwar <mohammad.anwar@yahoo.com>},
    VERSION_FROM       => 'lib/Games/Domino.pm',
    ABSTRACT_FROM      => 'lib/Games/Domino.pm',
    LICENSE            => 'artistic_2',
    EXE_FILES          => [ 'script/play-domino' ],
    MIN_PERL_VERSION   => 5.006,
    CONFIGURE_REQUIRES => {
        'ExtUtils::MakeMaker' => 0,
    },
    BUILD_REQUIRES   => {
        'Test::More' => 0,
    },
    PREREQ_PM => {
        'Attribute::Memoize'      => '1.01',
        'List::Util'              => '1.19',
        'Term::ANSIColor::Markup' => '0.06',
        'Moo'                     => '2.000000',
        'namespace::clean'        => '0.24',
        'overload'                => 0,
        'Type::Tiny'              => '1.000005',
        'Term::ReadKey'           => '2.33',
        'Term::Screen::Lite'      => '0.03',
        'MooX::Options'           => '4.023',
        'Moo::Role'               => '2.000000',
    },
    dist  => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
    clean => { FILES => 'Games-Domino-*' },
    (eval { ExtUtils::MakeMaker->VERSION(6.46) } ? (META_MERGE => {
        'meta-spec' => { version => 2 },
        provides    => {
            'Games::Domino'         => { file => 'lib/Games/Domino.pm'       , version => '0.29' },
            'Games::Domino::Player' => { file => 'lib/Games/Domino/Player.pm', version => '0.29' },
            'Games::Domino::Tile'   => { file => 'lib/Games/Domino/Tile.pm'  , version => '0.29' },
            'Games::Domino::Params' => { file => 'lib/Games/Domino/Params.pm', version => '0.29' },
        },
        resources => {
            repository => {
                type => 'git',
                url  => 'https://github.com/manwar/Games-Domino.git',
                web  => 'https://github.com/manwar/Games-Domino',
            },
        }})
     : ()
    ),
);
